### PR TITLE
Major rewrite of Inet Layer InterfaceIterator/InterfaceAddressIterator

### PR DESCRIPTION
**Pull request by jaylogue**
-- Rewrote InterfaceIterator and InterfaceAddressIterator classes to eliminate numerous bugs in both the sockets and LwIP implementations.  This work addresses all the bugs described in [issue#387](https://github.com/openweave/openweave-core/issues/387).

-- Eliminated InterfaceIteratorBasis class.  Since InterfaceIterator and InterfaceAddressIterator have distinct semantics, a common, non-virtual base class wasn’t appropriate.

-- Enhanced the LwIP implementation of both classes to provide thread-safety when run on multi-threaded platforms.

-- Extended InterfaceIterator/InterfaceAddressIterator with additional methods for querying interface state (e.g. IsUp()).

-- Cleaned up API documentation.  Added description of iterator behavior on multi-thread platforms, and in the face of spontaneous changes to the underlying network interfaces.

### CODE REVIEW

**Comment by jaylogue** [comment]
I'm going to merge this unless anyone has any objections.

**Comment by gerickson** [comment]
Missed this PR; will take a look now.

File: src/inet/InetInterface.cpp [hunk 1]
```diff
@@ -119,235 +154,760 @@ NL_DLL_EXPORT INET_ERROR InterfaceNameToId(const char *intfName, InterfaceId& in
...
 #if WEAVE_SYSTEM_CONFIG_USE_LWIP
-    curIntf = NULL;
+
+    // Lock LwIP stack
+    LOCK_TCPIP_CORE();
+
+    // Verify the previous netif is still on the list if netifs.  If so,
+    // advance to the next nextif.
+    struct netif * prevNetif = mCurNetif;
+    for (mCurNetif = netif_list; mCurNetif != NULL; mCurNetif = mCurNetif->next)
```
Comment (gerickson) [inline src/inet/InetInterface.cpp]: IIRC, using the newer NETIF_FOREACH, when available, as demonstrated in IPEndPoint.h allows for an easier single-netif optimization in LwIP.

Comment (jaylogue) [reply src/inet/InetInterface.cpp]: Thanks.  I wasn't aware of that pattern.  Code updated accordingly.

**Pull request merged**